### PR TITLE
task - document ci improvements and usage

### PR DESCRIPTION
**Pull request by dschom**
This commit documents design considerations for mono repos and CI pipelines, as well some developer guidelines around best practices given how our CI is setup.

### CODE REVIEW

File: docs/reference/continuous-integration-for-monorepos.md [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+
+![scripted test split](../assets/ci/scripted-test-split.png "image_tooltip")
+
+Finally, it’s good to know that integration tests are actually a hybrid approach. Here the tests are fanned out across two instances, but each workspace is run in parallel. We ended up doing this because the integration tests are quite a bit longer than unit tests, and also require more memory. After a bunch of trial and error, this hybrid approach seemed to be the most optimal.
+
+
+## In Summary
+These are the approaches we’ve taken to improve CI performance for a mono repo. There’s actually way more nuance that can be read about in this internal [document](https://docs.google.com/document/d/1WM4sLc0qQDzvq_0bX_nfZ1o2BM4M6YzmP33n36NB49Q/edit#heading=h.2jo3ou7ec55z) that is a retrospective on CI performance optimizations and some of the other issues that were encountered along the way. At a higher level this reference probably all that an engineer really needs to know.
```
Comment (clouserw) [inline docs/reference/continuous-integration-for-monorepos.md:107]: Share this document with Mozilla?

Comment (clouserw) [reply docs/reference/continuous-integration-for-monorepos.md:107]: missing "is" before probably

File: docs/reference/continuous-integration-for-monorepos.md [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+
+![scripted test split](../assets/ci/scripted-test-split.png "image_tooltip")
+
+Finally, it’s good to know that integration tests are actually a hybrid approach. Here the tests are fanned out across two instances, but each workspace is run in parallel. We ended up doing this because the integration tests are quite a bit longer than unit tests, and also require more memory. After a bunch of trial and error, this hybrid approach seemed to be the most optimal.
+
+
+## In Summary
+These are the approaches we’ve taken to improve CI performance for a mono repo. There’s actually way more nuance that can be read about in this internal [document](https://docs.google.com/document/d/1WM4sLc0qQDzvq_0bX_nfZ1o2BM4M6YzmP33n36NB49Q/edit#heading=h.2jo3ou7ec55z) that is a retrospective on CI performance optimizations and some of the other issues that were encountered along the way. At a higher level this reference probably all that an engineer really needs to know. 
+
+There is still room for more improvement, but for the pass this is where we are at. Ideas for future improvements are:
```
Comment (clouserw) [inline docs/reference/continuous-integration-for-monorepos.md:109]: s/pass/present/

**Review by clouserw — APPROVED** [review]

File: docs/how-tos/ci-guidelines.md [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+
+Integration tests are tests that:
+- Take longer to run, typically over 500ms.
+- Run with just infrastructure, i.e. we must execute a `yarn start infrastructure` is required prior to running tests)
+
+In order to distinguish between these kinds of tests, we have created a tagging convention. Our tests names need to be tagged with either #unit or #integration. These tags can exist at the test level (i.e. the ‘it’ block) or at the suite level (i.e. the ‘describe’ block). 
+
+In the future we may have better conventions for this such as file names or even folders, but for now, this was the least disruptive way to separate the two types of tests without disrupting the git history too much.
+
+Important: For jest tests we need to be explicit and add the #unit and #integration tag to each top level 'describe’ block. This is due to the fact that jest does not support an [‘invert’ option[](https://mochajs.org/api/mocha#invert) like mocha does, so we have not way run all tests except those with #integration. Therefore we must explicitly tag all tests.
```
[the remark this answers is not in the document]
Comment (dschom) [reply docs/how-tos/ci-guidelines.md:51]: No, I'll file a follow up for this. That's a great idea.

[the remark this answers is not in the document]
Comment (dschom) [reply docs/how-tos/ci-guidelines.md:51]: https://mozilla-hub.atlassian.net/browse/FXA-6641

**Pull request merged**